memcmp() out-of-bounds fixes

MaxKellermann opened the pull request.

Girgias commented:
This PR is doing too much and unrelated things, can you please split them up into related PRs.

Also IMHO adding and using a newly introduced API should be two different commits.

MaxKellermann commented:
@Girgias better?

Girgias commented on ext/oci8/oci8.c:
```diff
@@ -1096,9 +1096,7 @@ php_oci_connection *php_oci_do_connect_ex(char *username, int username_len, char
-									(ZSTR_LEN(tmp->hash_key) == ZSTR_LEN(hashed_details.s)) &&
-									(memcmp(ZSTR_VAL(tmp->hash_key), ZSTR_VAL(hashed_details.s),
-									 ZSTR_LEN(tmp->hash_key)) == 0)) {
+									zend_string_equal_content(tmp->hash_key, hashed_details.s)) {
```
Probably should be ``zend_string_equals()``

MaxKellermann replied:
I tried to guess the best names of my new functions based on existing function names - I'll amend my patch tomorrow adjusting to your suggestion.

MaxKellermann replied:
Oh, this was not about my choice of naming, but about my choice of replacement function.
I chose `zend_string_equal_content()` because it does exactly what the code it replaces did: compare lengths, and then `memcmp()` on the value:
https://github.com/php/php-src/blob/dd81dd92643e98ea61bdc9091a47c5cae0925a4b/Zend/zend_string.h#L353-L356
`zend_string_equals()`, on the other hand, additionally checks the `zend_string` pointers for equality:
https://github.com/php/php-src/blob/dd81dd92643e98ea61bdc9091a47c5cae0925a4b/Zend/zend_string.h#L358-L361
This is slightly different from the existing old code. No semantic change, because the outcome of the check is the same. But I wanted to be careful and not introduce actual code changes in this pure-refactoring patch.
Please confirm that you do want this change, and I'll amend the patch.

nikic replied:
zend_string_equal_content() is a low-level function that is used in some optimized codepaths that have performed a pointer equality check separately. Generally, code should always be using zend_string_equals() unless there's some special reason not to.

Girgias commented:
LGTM, but I don't know what our policy is about adding new APIs

MaxKellermann commented on Zend/zend_string.h:
```diff
@@ -339,6 +339,16 @@ static zend_always_inline void zend_string_release_ex(zend_string *s, bool persi
+static zend_always_inline bool zend_string_equals_cstr(const zend_string *s1, const char *s2)
+{
+	return zend_string_equals_str(s1, s2, strlen(s2));
+}
```
I called it `_cstr` because it accepts a C string (i.e. null-terminated). If I pass a `size_t`, it's not a C string.
I don't quite understand what you want me to do, but I did find the existing nomenclature confusing. From the name of a function, I couldn't deduce what kind of parameters it accepts. In this PR, I tried my best to adapt to whatever naming was there.

MaxKellermann commented on Zend/zend_string.h:
```diff
@@ -367,7 +377,20 @@ static zend_always_inline bool zend_string_equals(zend_string *s1, zend_string *
+	return ZSTR_LEN(str) >= prefix_length && !memcmp(ZSTR_VAL(str), prefix, prefix_length);
+}
+
+static zend_always_inline bool zend_string_starts_with_val(const zend_string *str, const zend_string *prefix)
```
I don't understand ..?

MaxKellermann commented:
Thanks @nikic and @Girgias for the reviews. I rebased and force-pushed with these changes:
- using `zend_string_equals()` as suggested, even though it does more than the existing code
- renamed `zend_string_equals_str()` to `zend_string_equals_cstr()` (even though that's not a C string), dropped the old `_cstr()` function (no shortcut which does `strlen()`)
- renamed `zend_string_starts_with_str()` to `zend_string_starts_with_cstr()`
- renamed `zend_string_starts_with_val()` to just `zend_string_starts_with()` (no suffix = takes `zend_string`?)

Does that meet your change requests?

iluuu1994 commented on Zend/zend_alloc.c:
```diff
@@ -69,6 +69,7 @@
 # include "win32/winutil.h"
 #endif
 
+#include <stdbool.h>
```
This is probably unused?

MaxKellermann replied:
True. This commit was already merged (the rest of it), and this is what stgit left of it ... I need to drop it.

iluuu1994 commented on Zend/zend_string.h:
```diff
@@ -367,7 +372,20 @@ static zend_always_inline bool zend_string_equals(zend_string *s1, zend_string *
 
 #define zend_string_equals_literal(str, literal) \
-	(ZSTR_LEN(str) == sizeof(literal)-1 && !memcmp(ZSTR_VAL(str), literal, sizeof(literal) - 1))
+	zend_string_equals_cstr(str, literal, sizeof(literal) - 1)
```
Could use `strlen`

MaxKellermann replied:
Changed.

iluuu1994 reviewed (approved):
LGTM

MaxKellermann commented:
Curious question - @iluuu1994 you folded all commits of this PR into one - why? I took extra care to submit separate commits for separate aspects, to make my work easier to read & review, and to allow bisecting in case of bugs found later. Folding all comits breaks that.
If single-commit PRs are preferred here at php-src, I could save the time for doing separating the commits.

iluuu1994 commented:
https://wiki.php.net/vcs/gitworkflow

> Additionally, the history of pull requests often requires cleanup. For most pull requests, all commits can be squashed into one.

Your commit history is clean but often that's not the case and it makes more sense to squash. I guess in this case either would've been fine.
